File: flutter/lib/src/serving/serve_models.dart
/// Wire models for the API serving responses (`/ads`, `/config`), wire v1.
///
/// Parsing is defensive + forward-compatible: unknown fields are ignored, missing
/// fields fall back to safe defaults, and extra positional ad slots (future
/// `weight`, `package`, ...) are tolerated. Never throws on a well-formed-but-newer
/// payload, so an old SDK keeps working when the server adds fields.
library;

import 'dart:math';

/// One ad in the rotation. Wire form is a bare positional tuple
/// `[id, image, store, fill, "T,B,L,R"]` (see WIRE.md); only `[id, image]` are
/// required and every slot after `store` is optional + additive (older SDKs ignore
/// them, so the wire stays non-breaking). `id` is required for beacon attribution.
class AdItem {
  const AdItem({
    required this.id,
    required this.image,
    this.store,
    this.fill = 'contain',
    this.edgeColors = const [],
  });

  final int id;
  final String image;
  final String? store;

  /// How to present a full-screen creative whose aspect does not match the device.
  /// `contain` (default) = letterbox/pillarbox with a solid edge-colour fill (no
  /// stretch). `cover`/`blur` = photographic/full-bleed, so the creative covers and
  /// the gap behind is a blurred enlargement. Unknown -> treated as contain.
  final String fill;

  /// Server-sampled edge colours `[top, bottom, left, right]` (hex), used to fill the
  /// contain-fit gaps. Empty when the server has not sampled the creative yet; the SDK
  /// then falls back to a neutral solid. Pixel sampling happens once at creative
  /// generation (server side), never on device.
  final List<String> edgeColors;

  /// True when the creative wants full-bleed cover + blur fill instead of contain.
  bool get isPhotographic => fill == 'cover' || fill == 'blur';

  /// Parse `[id, image, store, fill, "T,B,L,R"]`. Tolerates extra trailing slots so
  /// the wire can grow without breaking already-shipped SDKs.
  static AdItem? fromList(dynamic row) {
    if (row is! List || row.length < 2) return null;
    final id = row[0];
    final image = row[1];
    if (id is! int || image is! String) return null;
    final store = row.length > 2 && row[2] is String ? row[2] as String : null;
    final fill =
        row.length > 3 && row[3] is String && (row[3] as String).isNotEmpty ? row[3] as String : 'contain';
    final colors = row.length > 4 && row[4] is String ? _parseColors(row[4] as String) : const <String>[];
    return AdItem(id: id, image: image, store: store, fill: fill, edgeColors: colors);
  }

  // "T,B,L,R" -> 4 hex strings, or [] unless exactly four are present (all-or-none).
  static List<String> _parseColors(String s) {
    final parts = [for (final p in s.split(',')) p.trim()]..removeWhere((p) => p.isEmpty);
    return parts.length == 4 ? parts : const [];
  }
}

/// The two-tier `/ads` response: `a` = cross-promo (reserve + fallback), `o` =
/// own-studio last resort.
class AdBundle {
  const AdBundle({required this.ads, required this.own, this.nonce = ''});

  final List<AdItem> ads;
  final List<AdItem> own;

  /// Per-serve, single-use nonce from the server, echoed on the beacon so the server can
  /// reject replays. Opaque to the SDK. Empty when the server didn't send one.
  final String nonce;

  static const AdBundle empty = AdBundle(ads: [], own: []);

  factory AdBundle.fromJson(Map<String, dynamic> json) => AdBundle(
        ads: _items(json['a']),
        own: _items(json['o']),
        nonce: json['n'] is String ? json['n'] as String : '',
      );

  static List<AdItem> _items(dynamic x) {
    if (x is! List) return const [];
    return [
      for (final row in x)
        if (AdItem.fromList(row) case final item?) item,
    ];
  }
}

/// Per-app serving config from `/config/<package>`. Missing fields fall back to
/// the house-friendly defaults (reserve 1-of-4, fallback on, no own-ads).
class ServeConfig {
  const ServeConfig({
    required this.reserveShare,
    required this.reserveOneIn,
    required this.fallbackFill,
    required this.fillOwnAds,
    required this.houseCooldownSec,
    required this.maxPerSession,
    required this.ownAdsCooldownMin,
    this.bannerRotationSec = 0,
    this.rewardedSeconds = 0,
    this.adBadgeChance = 0,
    this.badgeUrl = '',
    this.bannerSdkRefresh = false,
  });

  final bool reserveShare;
  final int reserveOneIn;
  final bool fallbackFill;
  final bool fillOwnAds;
  final int houseCooldownSec;
  final int maxPerSession;
  final int ownAdsCooldownMin;

  /// Banner rotation/window unit, in seconds. 0 = unset -> the SDK jitters (~55-65s,
  /// avoids every device rotating in lockstep). Use [bannerRotation] to resolve it.
  final int bannerRotationSec;

  /// How long the rewarded countdown runs, in seconds. 0 = unset -> the SDK default
  /// (10s). Operator-set; the countdown starts after the creative loads. Use
  /// [rewardedDuration] to resolve it.
  final int rewardedSeconds;

  /// Probability (0..1) of drawing a small disclosure badge over a house ad ("Ad" on
  /// full-screen, a "GK" mark on banners). 0 = never, 1 = always; rolled per display.
  final double adBadgeChance;

  /// Where a tap on the full-screen badge goes. '' -> the SDK default (kBadgeInfoUrl).
  final String badgeUrl;

  /// Host's banner refresh strategy (portal Config). false = Regular (default): SDK shows
  /// one house unit then hands back to paid for N-1 units. true = Advanced: SDK drives the
  /// refresh (host must disable paid auto-refresh). The widget's explicit `sdkControlsRefresh`
  /// param overrides this; otherwise the banner defaults from here.
  final bool bannerSdkRefresh;

  /// Compiled fallback when no config has ever been fetched (offline first run).
  static const ServeConfig defaults = ServeConfig(
    reserveShare: true,
    reserveOneIn: 4,
    fallbackFill: true,
    fillOwnAds: false,
    houseCooldownSec: 240,
    maxPerSession: 3,
    ownAdsCooldownMin: 5,
    bannerRotationSec: 0,
    rewardedSeconds: 0,
    adBadgeChance: 0,
    badgeUrl: '',
    bannerSdkRefresh: false,
  );

  /// Effective banner rotation interval: the configured value if set (>0), else a
  /// jittered ~55-65s default (55 + 0..10). Call once per banner so the jitter is stable.
  Duration bannerRotation([Random? rng]) => bannerRotationSec > 0
      ? Duration(seconds: bannerRotationSec)
      : Duration(seconds: 55 + (rng ?? Random()).nextInt(11));

  /// Effective rewarded countdown: the configured value if set (>0), else the 10s default.
  Duration rewardedDuration() => Duration(seconds: rewardedSeconds > 0 ? rewardedSeconds : 10);

  /// Roll whether to draw the disclosure badge on this display (probability adBadgeChance).
  bool rollAdBadge([Random? rng]) =>
      adBadgeChance > 0 && (rng ?? Random()).nextDouble() < adBadgeChance;

  factory ServeConfig.fromJson(Map<String, dynamic> json) => ServeConfig(
        reserveShare: _b(json['reserve_share'], defaults.reserveShare),
        reserveOneIn: _i(json['reserve_one_in'], defaults.reserveOneIn),
        fallbackFill: _b(json['fallback_fill'], defaults.fallbackFill),
        fillOwnAds: _b(json['fill_own_ads'], defaults.fillOwnAds),
        houseCooldownSec: _i(json['house_cooldown_sec'], defaults.houseCooldownSec),
        maxPerSession: _i(json['max_per_session'], defaults.maxPerSession),
        ownAdsCooldownMin: _i(json['own_ads_cooldown_min'], defaults.ownAdsCooldownMin),
        bannerRotationSec: _i(json['banner_rotation_sec'], 0),
        rewardedSeconds: _i(json['rewarded_seconds'], 0),
        adBadgeChance: _d(json['ad_badge_chance'], 0),
        badgeUrl: json['badge_url'] is String ? json['badge_url'] as String : '',
        bannerSdkRefresh: _b(json['banner_sdk_refresh'], false),
      );

  static bool _b(dynamic v, bool d) => v is bool ? v : d;
  static int _i(dynamic v, int d) => v is int ? v : (v is num ? v.toInt() : d);
  static double _d(dynamic v, double d) => v is num ? v.toDouble() : d;
}

File: unity/Runtime/ServeModels.cs
using System;
using System.Collections.Generic;

namespace GoldenKrill
{
    /// One ad row: a bare positional tuple [id, image, store]. `store` is a /c tracker URL
    /// (open verbatim; it 302s to the store and counts the click), or a direct URL in test mode.
    public sealed class AdItem
    {
        public int Id;
        public string Image;
        public string Store; // may be null

        /// Parse a [id, image, store] row (tolerates extra trailing slots). Null if malformed.
        public static AdItem FromList(object row)
        {
            if (!(row is List<object> list) || list.Count < 2) return null;
            if (!(list[0] is double idD) || !(list[1] is string image)) return null;
            string store = list.Count > 2 && list[2] is string s ? s : null;
            return new AdItem { Id = (int)idD, Image = image, Store = store };
        }
    }

    /// The two-tier /ads response: Ads = cross-promo, Own = own-studio last resort. IsTest is
    /// the additive GK1 `t` tag (true only for a test-mode bundle); render the TEST AD normally.
    public sealed class AdBundle
    {
        public List<AdItem> Ads = new List<AdItem>();
        public List<AdItem> Own = new List<AdItem>();
        public string Nonce = "";
        public bool IsTest;

        public static readonly AdBundle Empty = new AdBundle();

        public static AdBundle FromJson(object json)
        {
            var map = json as Dictionary<string, object>;
            var b = new AdBundle();
            if (map == null) return b;
            b.Ads = Items(Get(map, "a"));
            b.Own = Items(Get(map, "o"));
            b.Nonce = Get(map, "n") is string n ? n : "";
            b.IsTest = Get(map, "t") is double t && t != 0;
            return b;
        }

        private static List<AdItem> Items(object x)
        {
            var outList = new List<AdItem>();
            if (x is List<object> list)
                foreach (var row in list)
                {
                    var it = AdItem.FromList(row);
                    if (it != null) outList.Add(it);
                }
            return outList;
        }

        internal static object Get(Dictionary<string, object> map, string key) =>
            map != null && map.TryGetValue(key, out var v) ? v : null;
    }

    /// Per-app serving config from /config. Missing fields fall back to house-friendly defaults.
    public sealed class ServeConfig
    {
        public bool ReserveShare;
        public int ReserveOneIn;
        public bool FallbackFill;
        public bool FillOwnAds;
        public int HouseCooldownSec;
        public int MaxPerSession;
        public int OwnAdsCooldownMin;
        public int BannerRotationSec; // 0 = SDK jitters ~55-65s
        public int RewardedSeconds;   // 0 = SDK default 10s
        public double AdBadgeChance;  // 0..1 probability of drawing the disclosure badge
        public string BadgeUrl;       // '' -> SDK default (Endpoints.BadgeInfoUrl)
        public bool BannerSdkRefresh; // false = Regular (passive), true = Advanced (SDK-driven)

        /// EXACT defaults, matching Flutter/RN/server (reserveOneIn=4, fallbackFill=true,
        /// houseCooldownSec=240, maxPerSession=3, fillOwnAds=false).
        public static ServeConfig Defaults => new ServeConfig
        {
            ReserveShare = true,
            ReserveOneIn = 4,
            FallbackFill = true,
            FillOwnAds = false,
            HouseCooldownSec = 240,
            MaxPerSession = 3,
            OwnAdsCooldownMin = 5,
            BannerRotationSec = 0,
            RewardedSeconds = 0,
            AdBadgeChance = 0,
            BadgeUrl = "",
            BannerSdkRefresh = false,
        };

        public static ServeConfig FromJson(object json)
        {
            var map = json as Dictionary<string, object>;
            var d = Defaults;
            if (map == null) return d;
            return new ServeConfig
            {
                ReserveShare = B(map, "reserve_share", d.ReserveShare),
                ReserveOneIn = I(map, "reserve_one_in", d.ReserveOneIn),
                FallbackFill = B(map, "fallback_fill", d.FallbackFill),
                FillOwnAds = B(map, "fill_own_ads", d.FillOwnAds),
                HouseCooldownSec = I(map, "house_cooldown_sec", d.HouseCooldownSec),
                MaxPerSession = I(map, "max_per_session", d.MaxPerSession),
                OwnAdsCooldownMin = I(map, "own_ads_cooldown_min", d.OwnAdsCooldownMin),
                BannerRotationSec = I(map, "banner_rotation_sec", 0),
                RewardedSeconds = I(map, "rewarded_seconds", 0),
                AdBadgeChance = AdBundle.Get(map, "ad_badge_chance") is double a ? a : 0,
                BadgeUrl = AdBundle.Get(map, "badge_url") is string s ? s : "",
                BannerSdkRefresh = B(map, "banner_sdk_refresh", d.BannerSdkRefresh),
            };
        }

        private static bool B(Dictionary<string, object> m, string k, bool d) =>
            AdBundle.Get(m, k) is bool v ? v : d;

        private static int I(Dictionary<string, object> m, string k, int d) =>
            AdBundle.Get(m, k) is double v ? (int)Math.Truncate(v) : d;

        // --- Pure helpers (parity with models.ts) ---

        /// Roll whether to draw the disclosure badge on this display (probability AdBadgeChance).
        public bool RollAdBadge(Func<double> rnd) =>
            AdBadgeChance > 0 && rnd() < AdBadgeChance;

        /// Effective banner rotation interval (ms): configured value if set, else jittered ~55-65s.
        public int BannerRotationMs(Func<double> rnd)
        {
            int sec = BannerRotationSec > 0 ? BannerRotationSec : 55 + (int)(rnd() * 11);
            return sec * 1000;
        }

        /// Effective rewarded countdown (ms): configured value if set, else the 10s default.
        public int RewardedMs() => (RewardedSeconds > 0 ? RewardedSeconds : 10) * 1000;
    }
}
